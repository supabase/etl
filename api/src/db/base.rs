use crate::encryption::{DecryptionError, EncryptionError, EncryptionKey};
use serde::de::DeserializeOwned;
use serde::{Deserialize, Serialize};
use thiserror::Error;

#[derive(Debug, Error)]
pub enum ToDbError {
    #[error("An encryption error happened while converting to the db variant: {0}")]
    Encryption(#[from] EncryptionError),
}

#[derive(Debug, Error)]
pub enum ToMemoryError {
    #[error("A decryption error happened while converting to the memory variant: {0}")]
    Decryption(#[from] DecryptionError),
}

pub trait ToDb<T> {
    fn to_db_type(self, encryption_key: &EncryptionKey) -> Result<T, ToDbError>;
}

impl<T> ToDb<T> for T {
    fn to_db_type(self, _: &EncryptionKey) -> Result<T, ToDbError> {
        Ok(self)
    }
}

pub trait ToMemory<T> {
    fn to_memory_type(self, encryption_key: &EncryptionKey) -> Result<T, ToMemoryError>;
}

impl<T> ToMemory<T> for T {
    fn to_memory_type(self, _: &EncryptionKey) -> Result<T, ToMemoryError> {
        Ok(self)
    }
}

#[derive(Debug, Error)]
pub enum DbSerializationError {
    #[error("Error while serializing data to the db: {0}")]
    Serde(#[from] serde_json::Error),

    #[error("An error occurred while converting to the db representation: {0}")]
    ToDb(#[from] ToDbError),
}

#[derive(Debug, Error)]
pub enum DbDeserializationError {
    #[error("Error while deserializing data from the db: {0}")]
    Serde(#[from] serde_json::Error),

    #[error("An error occurred while converting to the memory representation: {0}")]
    ToMemory(#[from] ToMemoryError),
}

pub fn serialize_to_db_as_json<T, S>(
    value: T,
    encryption_key: &EncryptionKey,
) -> Result<serde_json::Value, DbSerializationError>
where
    T: ToDb<S>,
    S: Serialize,
{
    let value = value.to_db_type(encryption_key)?;
    let serialized_value = serde_json::to_value(value)?;

    Ok(serialized_value)
}

pub fn deserialize_from_db_json_str<T, S>(
    value: &str,
    encryption_key: &EncryptionKey,
) -> Result<S, DbDeserializationError>
where
    T: ToMemory<S>,
    T: DeserializeOwned,
{
    let deserialized_value: T = serde_json::from_str(value)?;
    let value = deserialized_value.to_memory_type(encryption_key)?;

    Ok(value)
}

pub fn deserialize_from_db_json_value<T, S>(
    value: serde_json::Value,
    encryption_key: &EncryptionKey,
) -> Result<S, DbDeserializationError>
where
    T: ToMemory<S>,
    T: DeserializeOwned,
{
    let deserialized_value: T = serde_json::from_value(value)?;
    let value = deserialized_value.to_memory_type(encryption_key)?;

    Ok(value)
}
