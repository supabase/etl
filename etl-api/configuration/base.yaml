supabase_api:
  url: "http://localhost:8080"
