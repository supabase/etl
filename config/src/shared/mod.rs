mod base;
mod connection;
mod destination;
mod pipeline;
mod replicator;
mod supabase;

pub use base::*;
pub use connection::*;
pub use destination::*;
pub use pipeline::*;
pub use replicator::*;
pub use supabase::*;
