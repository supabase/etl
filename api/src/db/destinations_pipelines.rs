use aws_lc_rs::error::Unspecified;
use sqlx::PgPool;
use thiserror::Error;

use crate::encryption::EncryptionKey;

use super::{
    destinations::{
        create_destination_txn, update_destination_txn, DestinationsDbError,
    },
    pipelines::{create_pipeline_txn, update_pipeline_txn},
};

#[derive(Debug, Error)]
pub enum DestinationPipelineDbError {
    #[error("sqlx error: {0}")]
    Sqlx(#[from] sqlx::Error),

    #[error("encryption error: {0}")]
    Encryption(#[from] Unspecified),

    #[error("sources error: {0}")]
    Destinations(#[from] DestinationsDbError),

    #[error("destination with id {0} not found")]
    DestinationNotFound(i64),

    #[error("pipeline with id {0} not found")]
    PipelineNotFound(i64),
}

#[expect(clippy::too_many_arguments)]
pub async fn create_destination_and_pipeline(
    pool: &PgPool,
    tenant_id: &str,
    source_id: i64,
    destination_name: &str,
    destination_config: DestinationConfig,
    image_id: i64,
    publication_name: &str,
    pipeline_config: PipelineConfig,
    encryption_key: &EncryptionKey,
) -> Result<(i64, i64), DestinationPipelineDbError> {
    let destination_config = destination_config.into_db_config(encryption_key)?;
    let destination_config =
        serde_json::to_value(destination_config).expect("failed to serialize config");
    let pipeline_config =
        serde_json::to_value(pipeline_config).expect("failed to serialize config");
    let mut txn = pool.begin().await?;
    let destination_id =
        create_destination_txn(&mut txn, tenant_id, destination_name, destination_config).await?;
    let pipeline_id = create_pipeline_txn(
        &mut txn,
        tenant_id,
        source_id,
        destination_id,
        image_id,
        publication_name,
        pipeline_config,
    )
    .await?;
    txn.commit().await?;
    Ok((destination_id, pipeline_id))
}

#[expect(clippy::too_many_arguments)]
pub async fn update_destination_and_pipeline(
    pool: &PgPool,
    tenant_id: &str,
    destination_id: i64,
    pipeline_id: i64,
    source_id: i64,
    destination_name: &str,
    destination_config: DestinationConfig,
    publication_name: &str,
    pipeline_config: PipelineConfig,
    encryption_key: &EncryptionKey,
) -> Result<(), DestinationPipelineDbError> {
    let destination_config = destination_config.into_db_config(encryption_key)?;
    let destination_config =
        serde_json::to_value(destination_config).expect("failed to serialize config");
    let pipeline_config =
        serde_json::to_value(pipeline_config).expect("failed to serialize config");
    let mut txn = pool.begin().await?;
    let destination_id_res = update_destination_txn(
        &mut txn,
        tenant_id,
        destination_name,
        destination_id,
        destination_config,
    )
    .await?;
    if destination_id_res.is_none() {
        txn.rollback().await?;
        return Err(DestinationPipelineDbError::DestinationNotFound(
            destination_id,
        ));
    };
    let pipeline_id_res = update_pipeline_txn(
        &mut txn,
        tenant_id,
        pipeline_id,
        source_id,
        destination_id,
        publication_name,
        pipeline_config,
    )
    .await?;

    if pipeline_id_res.is_none() {
        txn.rollback().await?;
        return Err(DestinationPipelineDbError::PipelineNotFound(pipeline_id));
    };

    txn.commit().await?;

    Ok(())
}
