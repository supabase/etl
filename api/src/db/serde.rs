use crate::encryption::{
    Decryptable, DecryptionError, Encryptable, EncryptionError, EncryptionKey,
};
use serde::de::DeserializeOwned;
use serde::Serialize;
use thiserror::Error;

#[derive(Debug, Error)]
pub enum DbSerializationError {
    #[error("Error while serializing data to the db: {0}")]
    Serde(#[from] serde_json::Error),

    #[error("An error occurred while encrypting data for the db representation: {0}")]
    Encryption(#[from] EncryptionError),
}

#[derive(Debug, Error)]
pub enum DbDeserializationError {
    #[error("Error while deserializing data from the db: {0}")]
    Serde(#[from] serde_json::Error),

    #[error("An error occurred while decrypting data from the db representation: {0}")]
    Decryption(#[from] DecryptionError),
}

pub fn serialize<S>(value: S) -> Result<serde_json::Value, DbSerializationError>
where
    S: Serialize,
{
    let serialized_value = serde_json::to_value(value)?;

    Ok(serialized_value)
}

pub fn encrypt_and_serialize<T, S>(
    value: T,
    encryption_key: &EncryptionKey,
) -> Result<serde_json::Value, DbSerializationError>
where
    T: Encryptable<S>,
    S: Serialize,
{
    let value = value.encrypt(encryption_key)?;
    let serialized_value = serde_json::to_value(value)?;

    Ok(serialized_value)
}

pub fn deserialize_from_value<S>(value: serde_json::Value) -> Result<S, DbDeserializationError>
where
    S: DeserializeOwned,
{
    let deserialized_value = serde_json::from_value(value)?;

    Ok(deserialized_value)
}

pub fn decrypt_and_deserialize_from_value<T, S>(
    value: serde_json::Value,
    encryption_key: &EncryptionKey,
) -> Result<S, DbDeserializationError>
where
    T: Decryptable<S>,
    T: DeserializeOwned,
{
    let deserialized_value: T = serde_json::from_value(value)?;
    let value = deserialized_value.decrypt(encryption_key)?;

    Ok(value)
}
