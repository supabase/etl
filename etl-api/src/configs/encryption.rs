use std::string;

use aws_lc_rs::{
    aead::{AES_256_GCM, Aad, Nonce, RandomizedNonceKey},
    rand::fill,
};
use base64::{Engine, prelude::BASE64_STANDARD};
use serde::{Deserialize, Serialize};
use thiserror::Error;

/// Errors that occur during data encryption.
#[derive(Debug, Error)]
pub enum EncryptionError {
    /// An unspecified error occurred while encrypting data.
    #[error("An unspecified error occurred while encrypting data")]
    Unspecified(#[from] aws_lc_rs::error::Unspecified),
}

/// Errors that occur during data decryption.
#[derive(Debug, Error)]
pub enum DecryptionError {
    /// An unspecified error occurred while decrypting data.
    #[error("An unspecified error occurred while decrypting data")]
    Unspecified(#[from] aws_lc_rs::error::Unspecified),

    /// Failed to decode base64 data during decryption.
    #[error("An error occurred while decoding BASE64 data for decryption: {0}")]
    Decode(#[from] base64::DecodeError),

    /// Failed to convert decrypted bytes to UTF-8 string.
    #[error("An error occurred while converting bytes to UTF-8 for decryption: {0}")]
    FromUtf8(#[from] string::FromUtf8Error),

    /// The key ID in the encrypted data did not match the expected key ID.
    #[error("There was a mismatch in the key id while decrypting data (got: {0}, expected: {1})")]
    MismatchedKeyId(u32, u32),
}

/// Trait for types that can be encrypted.
pub trait Encrypt<T> {
    /// Encrypts this value using the provided encryption key.
    fn encrypt(self, encryption_key: &EncryptionKey) -> Result<T, EncryptionError>;
}

/// Trait for types that can be decrypted.
pub trait Decrypt<T> {
    /// Decrypts this value using the provided encryption key.
    fn decrypt(self, encryption_key: &EncryptionKey) -> Result<T, DecryptionError>;
}

/// Encryption key with identifier for key management.
pub struct EncryptionKey {
    /// Unique identifier for the key.
    pub id: u32,
    /// The key material used for encryption and decryption.
    pub key: RandomizedNonceKey,
}

/// Encrypted value with metadata for decryption.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct EncryptedValue {
    /// Identifier of the key used for encryption.
    pub id: u32,
    /// Base64-encoded nonce used during encryption.
    pub nonce: String,
    /// Base64-encoded encrypted value.
    pub value: String,
}

/// Encrypts a string using AES-256-GCM encryption.
///
/// Returns an [`EncryptedValue`] containing the key ID, nonce, and encrypted data,
/// all base64-encoded for safe storage and transmission.
pub fn encrypt_text(
    value: String,
    encryption_key: &EncryptionKey,
) -> Result<EncryptedValue, EncryptionError> {
    let (encrypted_password, nonce) = encrypt(value.as_bytes(), &encryption_key.key)?;
    let encoded_encrypted_password = BASE64_STANDARD.encode(encrypted_password);
    let encoded_nonce = BASE64_STANDARD.encode(nonce.as_ref());

    Ok(EncryptedValue {
        id: encryption_key.id,
        nonce: encoded_nonce,
        value: encoded_encrypted_password,
    })
}

/// Decrypts an [`EncryptedValue`] back to the original string.
///
/// Validates the key ID matches before attempting decryption. Returns the original
/// plaintext string if decryption succeeds.
pub fn decrypt_text(
    encrypted_value: EncryptedValue,
    encryption_key: &EncryptionKey,
) -> Result<String, DecryptionError> {
    if encrypted_value.id != encryption_key.id {
        return Err(DecryptionError::MismatchedKeyId(
            encrypted_value.id,
            encryption_key.id,
        ));
    }

    let encrypted_value_bytes = BASE64_STANDARD.decode(encrypted_value.value)?;
    let nonce = Nonce::try_assume_unique_for_key(&BASE64_STANDARD.decode(encrypted_value.nonce)?)?;

    let decrypted_value_bytes = decrypt(encrypted_value_bytes, nonce, &encryption_key.key)?;

    let decrypted_value = String::from_utf8(decrypted_value_bytes)?;

    Ok(decrypted_value)
}

/// Encrypts bytes using AES-256-GCM with a randomized nonce.
///
/// Returns the encrypted data and the nonce used for encryption.
fn encrypt(
    plaintext: &[u8],
    key: &RandomizedNonceKey,
) -> Result<(Vec<u8>, Nonce), aws_lc_rs::error::Unspecified> {
    let mut in_out = plaintext.to_vec();
    let nonce = key.seal_in_place_append_tag(Aad::empty(), &mut in_out)?;

    Ok((in_out, nonce))
}

/// Decrypts AES-256-GCM encrypted data using the key and nonce.
///
/// Returns the original plaintext bytes.
fn decrypt(
    mut ciphertext: Vec<u8>,
    nonce: Nonce,
    key: &RandomizedNonceKey,
) -> Result<Vec<u8>, aws_lc_rs::error::Unspecified> {
    let plaintext = key.open_in_place(nonce, Aad::empty(), &mut ciphertext)?;

    Ok(plaintext.to_vec())
}

/// Generates a cryptographically secure random encryption key.
///
/// Creates a new [`RandomizedNonceKey`] for AES-256-GCM encryption using
/// secure random bytes.
///
/// # Panics
/// Panics if `T` doesn't match the required key length for AES-256-GCM.
pub fn generate_random_key<const T: usize>()
-> Result<RandomizedNonceKey, aws_lc_rs::error::Unspecified> {
    let mut key_bytes = [0u8; T];
    fill(&mut key_bytes)?;

    let key = RandomizedNonceKey::new(&AES_256_GCM, &key_bytes)?;

    Ok(key)
}
