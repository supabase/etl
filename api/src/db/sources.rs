use postgres::sqlx::config::PgConnectionConfig;
use secrecy::Secret;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sqlx::{PgPool, Postgres, Transaction};
use std::fmt::Debug;
use thiserror::Error;

use crate::db::base::{
    deserialize_from_db_json_value, serialize_to_db_as_json, DbDeserializationError,
    DbSerializationError, ToDb, ToDbError, ToMemory, ToMemoryError,
};
use crate::encryption::{decrypt_text, encrypt_text, EncryptedValue, EncryptionKey};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub struct SourceConfig {
    host: String,
    port: u16,
    name: String,
    username: String,
    password: Option<String>,
}

impl SourceConfig {
    pub fn into_connection_config(self) -> PgConnectionConfig {
        PgConnectionConfig {
            host: self.host,
            port: self.port,
            name: self.name,
            username: self.username,
            password: self.password.map(Secret::new),
            // By default, we enable ssl.
            require_ssl: true,
        }
    }
}

impl ToDb<SourceConfigInDb> for SourceConfig {
    fn to_db_type(self, encryption_key: &EncryptionKey) -> Result<SourceConfigInDb, ToDbError> {
        let mut encrypted_password = None;
        if let Some(password) = self.password {
            encrypted_password = Some(encrypt_text(password, encryption_key)?);
        }

        Ok(SourceConfigInDb {
            host: self.host,
            port: self.port,
            name: self.name,
            username: self.username,
            password: encrypted_password,
        })
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub struct SourceConfigInDb {
    host: String,
    port: u16,
    name: String,
    username: String,
    password: Option<EncryptedValue>,
}

impl ToMemory<SourceConfig> for SourceConfigInDb {
    fn to_memory_type(self, encryption_key: &EncryptionKey) -> Result<SourceConfig, ToMemoryError> {
        let mut decrypted_password = None;
        if let Some(password) = self.password {
            decrypted_password = Some(decrypt_text(password, encryption_key)?);
        }

        Ok(SourceConfig {
            host: self.host,
            port: self.port,
            name: self.name,
            username: self.username,
            password: decrypted_password,
        })
    }
}

#[derive(Debug)]
pub struct Source {
    pub id: i64,
    pub tenant_id: String,
    pub name: String,
    pub config: SourceConfig,
}

#[derive(Debug, Error)]
pub enum SourcesDbError {
    #[error("Error while dealing with PostgreSQL: {0}")]
    Sqlx(#[from] sqlx::Error),

    #[error("Error while serializing source config: {0}")]
    DbSerializationError(#[from] DbSerializationError),

    #[error("Error while deserializing source config: {0}")]
    DbDeserializationError(#[from] DbDeserializationError),
}

pub async fn create_source(
    pool: &PgPool,
    tenant_id: &str,
    name: &str,
    config: SourceConfig,
    encryption_key: &EncryptionKey,
) -> Result<i64, SourcesDbError> {
    let config = serialize_to_db_as_json::<SourceConfig, SourceConfigInDb>(config, encryption_key)?;

    let mut txn = pool.begin().await?;
    let res = create_source_txn(&mut txn, tenant_id, name, config).await;
    txn.commit().await?;

    res
}

pub async fn create_source_txn(
    txn: &mut Transaction<'_, Postgres>,
    tenant_id: &str,
    name: &str,
    config: Value,
) -> Result<i64, SourcesDbError> {
    let record = sqlx::query!(
        r#"
        insert into app.sources (tenant_id, name, config)
        values ($1, $2, $3)
        returning id
        "#,
        tenant_id,
        name,
        config
    )
    .fetch_one(&mut **txn)
    .await?;

    Ok(record.id)
}

pub async fn read_source(
    pool: &PgPool,
    tenant_id: &str,
    source_id: i64,
    encryption_key: &EncryptionKey,
) -> Result<Option<Source>, SourcesDbError> {
    let record = sqlx::query!(
        r#"
        select id, tenant_id, name, config
        from app.sources
        where tenant_id = $1 and id = $2
        "#,
        tenant_id,
        source_id,
    )
    .fetch_optional(pool)
    .await?;

    let source = record
        .map(|r| {
            let config = deserialize_from_db_json_value::<SourceConfigInDb, SourceConfig>(
                r.config,
                encryption_key,
            )?;
            let source = Source {
                id: r.id,
                tenant_id: r.tenant_id,
                name: r.name,
                config,
            };
            Ok::<Source, SourcesDbError>(source)
        })
        .transpose()?;
    Ok(source)
}

pub async fn update_source(
    pool: &PgPool,
    tenant_id: &str,
    name: &str,
    source_id: i64,
    config: SourceConfig,
    encryption_key: &EncryptionKey,
) -> Result<Option<i64>, SourcesDbError> {
    let config = serialize_to_db_as_json::<SourceConfig, SourceConfigInDb>(config, encryption_key)?;

    let record = sqlx::query!(
        r#"
        update app.sources
        set config = $1, name = $2
        where tenant_id = $3 and id = $4
        returning id
        "#,
        config,
        name,
        tenant_id,
        source_id
    )
    .fetch_optional(pool)
    .await?;

    Ok(record.map(|r| r.id))
}

pub async fn delete_source(
    pool: &PgPool,
    tenant_id: &str,
    source_id: i64,
) -> Result<Option<i64>, sqlx::Error> {
    let record = sqlx::query!(
        r#"
        delete from app.sources
        where tenant_id = $1 and id = $2
        returning id
        "#,
        tenant_id,
        source_id
    )
    .fetch_optional(pool)
    .await?;

    Ok(record.map(|r| r.id))
}

pub async fn read_all_sources(
    pool: &PgPool,
    tenant_id: &str,
    encryption_key: &EncryptionKey,
) -> Result<Vec<Source>, SourcesDbError> {
    let records = sqlx::query!(
        r#"
        select id, tenant_id, name, config
        from app.sources
        where tenant_id = $1
        "#,
        tenant_id,
    )
    .fetch_all(pool)
    .await?;

    let mut sources = Vec::with_capacity(records.len());
    for record in records {
        let config = deserialize_from_db_json_value::<SourceConfigInDb, SourceConfig>(
            record.config.clone(),
            encryption_key,
        )?;
        let source = Source {
            id: record.id,
            tenant_id: record.tenant_id,
            name: record.name,
            config,
        };
        sources.push(source);
    }

    Ok(sources)
}

pub async fn source_exists(
    pool: &PgPool,
    tenant_id: &str,
    source_id: i64,
) -> Result<bool, sqlx::Error> {
    let record = sqlx::query!(
        r#"
        select exists (select id
        from app.sources
        where tenant_id = $1 and id = $2) as "exists!"
        "#,
        tenant_id,
        source_id
    )
    .fetch_one(pool)
    .await?;

    Ok(record.exists)
}

#[cfg(test)]
mod tests {
    use crate::db::sources::{SourceConfig, SourceConfigInDb};
    use crate::encryption::EncryptionKey;
    use aws_lc_rs::aead::RandomizedNonceKey;
    use serde_json;

    #[test]
    pub fn source_config_json_roundtrip() {
        let json = r#"{
            "host": "localhost",
            "port": 5432,
            "name": "postgres",
            "username": "postgres",
            "password": "postgres",
        }"#;
        let expected = SourceConfig {
            host: "localhost".to_string(),
            port: 5432,
            name: "postgres".to_string(),
            username: "postgres".to_string(),
            password: Some("postgres".to_string()),
        };

        let deserialized = serde_json::from_str::<SourceConfig>(json);
        assert!(deserialized.is_ok());
        assert_eq!(expected, deserialized.as_ref().unwrap().to_owned());

        let serialized = serde_json::to_string_pretty(&expected);
        assert!(serialized.is_ok());
        assert_eq!(json, serialized.unwrap());
    }

    #[test]
    pub fn source_config_in_db_password_encryption() {
        use crate::db::base::{deserialize_from_db_json_value, serialize_to_db_as_json};
        let key_bytes = [42u8; 32];
        let key = RandomizedNonceKey::new(&aws_lc_rs::aead::AES_256_GCM, &key_bytes).unwrap();
        let encryption_key = EncryptionKey { id: 1, key };

        let config = SourceConfig {
            host: "localhost".to_string(),
            port: 5432,
            name: "postgres".to_string(),
            username: "postgres".to_string(),
            password: Some("supersecret".to_string()),
        };

        // Serialize to db (should encrypt password)
        let config_in_db = serialize_to_db_as_json::<SourceConfig, SourceConfigInDb>(
            config.clone(),
            &encryption_key,
        )
        .unwrap();
        // Deserialize from db (should decrypt password)
        let deserialized_config = deserialize_from_db_json_value::<SourceConfigInDb, SourceConfig>(
            config_in_db,
            &encryption_key,
        )
        .unwrap();
        assert_eq!(config, deserialized_config);

        // None password case
        let config = SourceConfig {
            password: None,
            ..config.clone()
        };
        let config_in_db = serialize_to_db_as_json::<SourceConfig, SourceConfigInDb>(
            config.clone(),
            &encryption_key,
        )
        .unwrap();
        let deserialized_config = deserialize_from_db_json_value::<SourceConfigInDb, SourceConfig>(
            config_in_db,
            &encryption_key,
        )
        .unwrap();
        assert_eq!(config, deserialized_config);
    }
}
