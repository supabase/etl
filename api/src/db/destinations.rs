use sqlx::{PgPool, Postgres, Transaction};
use std::{
    fmt::{Debug, Formatter},
    str::{from_utf8, Utf8Error},
};
use thiserror::Error;
use config::shared::DestinationConfig;
use crate::db::base::{decrypt_and_deserialize_from_value, encrypt_and_serialize, DbDeserializationError, DbSerializationError, Decryptable, Encryptable, ToDbError, ToMemoryError};
use crate::db::sources::EncryptedSourceConfig;
use crate::encryption::{decrypt, encrypt, encrypt_text, EncryptedValue, EncryptionKey};

impl Encryptable<EncryptedDestinationConfig> for DestinationConfig {
    
    fn encrypt(self, encryption_key: &EncryptionKey) -> Result<EncryptedDestinationConfig, ToDbError> {
        match self {
            Self::BigQuery {
                project_id,
                dataset_id,
                service_account_key,
                max_staleness_mins
            } => {
                let mut encrypted_service_account_key = encrypt_text(service_account_key, encryption_key)?;
                
                Ok(EncryptedDestinationConfig::BigQuery {
                    project_id,
                    dataset_id,
                    service_account_key: encrypted_service_account_key,
                    max_staleness_mins,
                })
            }
            _ => {}
        }
    }
}

#[derive(serde::Serialize, serde::Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum EncryptedDestinationConfig {
    BigQuery {
        project_id: String,
        dataset_id: String,
        service_account_key: EncryptedValue,
        #[serde(skip_serializing_if = "Option::is_none")]
        max_staleness_mins: Option<u16>,
    },
}

impl Decryptable<DestinationConfig> for EncryptedDestinationConfig {
    
    fn decrypt(self, encryption_key: &EncryptionKey) -> Result<DestinationConfig, ToMemoryError> {
        todo!()
    }
}

#[derive(Debug, Error)]
pub enum DestinationsDbError {
    #[error("Error while interacting with PostgreSQL for destinations: {0}")]
    Sqlx(#[from] sqlx::Error),

    #[error("Error while serializing destination config: {0}")]
    DbSerializationError(#[from] DbSerializationError),

    #[error("Error while deserializing destination config: {0}")]
    DbDeserializationError(#[from] DbDeserializationError),
}

pub struct Destination {
    pub id: i64,
    pub tenant_id: String,
    pub name: String,
    pub config: DestinationConfig,
}

pub async fn create_destination(
    pool: &PgPool,
    tenant_id: &str,
    name: &str,
    config: DestinationConfig,
    encryption_key: &EncryptionKey,
) -> Result<i64, DestinationsDbError> {
    let config = encrypt_and_serialize(config, encryption_key)?;
    
    let mut txn = pool.begin().await?;
    let res = create_destination_txn(&mut txn, tenant_id, name, config).await;
    txn.commit().await?;
    
    res
}

pub async fn create_destination_txn(
    txn: &mut Transaction<'_, Postgres>,
    tenant_id: &str,
    name: &str,
    config: serde_json::Value,
) -> Result<i64, DestinationsDbError> {
    let record = sqlx::query!(
        r#"
        insert into app.destinations (tenant_id, name, config)
        values ($1, $2, $3)
        returning id
        "#,
        tenant_id,
        name,
        config
    )
    .fetch_one(&mut **txn)
    .await?;

    Ok(record.id)
}

pub async fn read_destination(
    pool: &PgPool,
    tenant_id: &str,
    destination_id: i64,
    encryption_key: &EncryptionKey,
) -> Result<Option<Destination>, DestinationsDbError> {
    let record = sqlx::query!(
        r#"
        select id, tenant_id, name, config
        from app.destinations
        where tenant_id = $1 and id = $2
        "#,
        tenant_id,
        destination_id,
    )
    .fetch_optional(pool)
    .await?;

    let destination = match record {
        Some(record) => { 
            let config = decrypt_and_deserialize_from_value::<EncryptedDestinationConfig, DestinationConfig>(record.config, encryption_key)?;
            
            let destination = Destination {
                id: record.id,
                tenant_id: record.tenant_id,
                name: record.name,
                config,
            };
            
            Some(destination)
        },
        None => None
    };
    
    Ok(destination)
}

pub async fn update_destination(
    pool: &PgPool,
    tenant_id: &str,
    name: &str,
    destination_id: i64,
    config: DestinationConfig,
    encryption_key: &EncryptionKey,
) -> Result<Option<i64>, DestinationsDbError> {
    let config = encrypt_and_serialize(config, encryption_key)?;
   
    let mut txn = pool.begin().await?;
    let res = update_destination_txn(
        &mut txn,
        tenant_id,
        name,
        destination_id,
        config,
    )
    .await;
    txn.commit().await?;
    
    res
}

pub async fn update_destination_txn(
    txn: &mut Transaction<'_, Postgres>,
    tenant_id: &str,
    name: &str,
    destination_id: i64,
    config: serde_json::Value,
) -> Result<Option<i64>, DestinationsDbError> {
    let record = sqlx::query!(
        r#"
        update app.destinations
        set config = $1, name = $2
        where tenant_id = $3 and id = $4
        returning id
        "#,
        config,
        name,
        tenant_id,
        destination_id
    )
    .fetch_optional(&mut **txn)
    .await?;

    Ok(record.map(|r| r.id))
}

pub async fn delete_destination(
    pool: &PgPool,
    tenant_id: &str,
    destination_id: i64,
) -> Result<Option<i64>, sqlx::Error> {
    let record = sqlx::query!(
        r#"
        delete from app.destinations
        where tenant_id = $1 and id = $2
        returning id
        "#,
        tenant_id,
        destination_id
    )
    .fetch_optional(pool)
    .await?;

    Ok(record.map(|r| r.id))
}

pub async fn read_all_destinations(
    pool: &PgPool,
    tenant_id: &str,
    encryption_key: &EncryptionKey,
) -> Result<Vec<Destination>, DestinationsDbError> {
    let records = sqlx::query!(
        r#"
        select id, tenant_id, name, config
        from app.destinations
        where tenant_id = $1
        "#,
        tenant_id,
    )
    .fetch_all(pool)
    .await?;

    let mut destinations = Vec::with_capacity(records.len());
    for record in records {
        let config: DestinationConfigInDb = serde_json::from_value(record.config)?;
        let config = config.into_config(encryption_key)?;
        let source = Destination {
            id: record.id,
            tenant_id: record.tenant_id,
            name: record.name,
            config,
        };
        destinations.push(source);
    }

    Ok(destinations)
}

pub async fn destination_exists(
    pool: &PgPool,
    tenant_id: &str,
    destination_id: i64,
) -> Result<bool, sqlx::Error> {
    let record = sqlx::query!(
        r#"
        select exists (select id
        from app.destinations
        where tenant_id = $1 and id = $2) as "exists!"
        "#,
        tenant_id,
        destination_id,
    )
    .fetch_one(pool)
    .await?;

    Ok(record.exists)
}

// TODO: rewrite tests.
#[cfg(test)]
mod tests {
    use crate::db::destinations::DestinationConfig;

    #[test]
    pub fn deserialize_settings_test() {
        let settings = r#"{
            "big_query": {
                "project_id": "project-id",
                "dataset_id": "dataset-id",
                "service_account_key": "service-account-key"
            }
        }"#;
        let actual = serde_json::from_str::<DestinationConfig>(settings);
        let expected = DestinationConfig::BigQuery {
            project_id: "project-id".to_string(),
            dataset_id: "dataset-id".to_string(),
            service_account_key: "service-account-key".to_string(),
            max_staleness_mins: None,
        };
        assert!(actual.is_ok());
        assert_eq!(expected, actual.unwrap());
    }

    #[test]
    pub fn serialize_settings_test() {
        let actual = DestinationConfig::BigQuery {
            project_id: "project-id".to_string(),
            dataset_id: "dataset-id".to_string(),
            service_account_key: "service-account-key".to_string(),
            max_staleness_mins: None,
        };
        let expected = r#"{"big_query":{"project_id":"project-id","dataset_id":"dataset-id","service_account_key":"service-account-key"}}"#;
        let actual = serde_json::to_string(&actual);
        assert!(actual.is_ok());
        assert_eq!(expected, actual.unwrap());
    }
}
