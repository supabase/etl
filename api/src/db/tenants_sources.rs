use sqlx::PgTransaction;
use std::ops::DerefMut;
use thiserror::Error;

use crate::db::serde::DbSerializationError;
use crate::db::sources::{SourceConfig, SourcesDbError, create_source};
use crate::db::tenants::{TenantsDbError, create_tenant};
use crate::encryption::EncryptionKey;

#[derive(Debug, Error)]
pub enum TenantSourceDbError {
    #[error("Error while interacting with PostgreSQL for tenants and/or sources: {0}")]
    Database(#[from] sqlx::Error),

    #[error("Error while serializing tenant or source config: {0}")]
    DbSerialization(#[from] DbSerializationError),

    #[error(transparent)]
    Sources(#[from] SourcesDbError),

    #[error(transparent)]
    Tenants(#[from] TenantsDbError),
}

pub async fn create_tenant_and_source(
    txn: &mut PgTransaction<'_>,
    tenant_id: &str,
    tenant_name: &str,
    source_name: &str,
    source_config: SourceConfig,
    encryption_key: &EncryptionKey,
) -> Result<(String, i64), TenantSourceDbError> {
    let tenant_id = create_tenant(txn.deref_mut(), tenant_id, tenant_name).await?;
    let source_id = create_source(
        txn.deref_mut(),
        &tenant_id,
        source_name,
        source_config,
        encryption_key,
    )
    .await?;

    Ok((tenant_id, source_id))
}
