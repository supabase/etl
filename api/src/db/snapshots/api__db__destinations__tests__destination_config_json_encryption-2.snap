---
source: api/src/db/destinations.rs
expression: deserialized_config
---
BigQuery {
    project_id: "project-id",
    dataset_id: "dataset-id",
    service_account_key: SerializableSecretString(
        "[REDACTED]",
    ),
    max_staleness_mins: Some(
        99,
    ),
}
